lib/ukboot/boot.c

```diff
@@ -377,12 +392,65 @@ void ukplat_entry(int argc, char *argv[])
 	print_banner(stdout);
 	fflush(stdout);
 
+#if !CONFIG_LIBUKBOOT_MAINTHREAD
+	do_main(ictx.cmdline.argc, ictx.cmdline.argv);
+	tctx.target = UKPLAT_HALT;
+
+#else /* CONFIG_LIBUKBOOT_MAINTHREAD */
+	m = uk_sched_thread_create_fn2(s, main_thread,
+				       (void *)((long)argc), (void *)argv,
+				       0x0 /* default stack size */,
+				       false, false,
+				       "main", NULL,
+				       main_thread_dtor);
+	if (unlikely(!m || PTRISERR(m))) {
+		uk_pr_err("Failed to launch application's main()\n");
+		goto exit;
+	}
+
+	/* Block execution of "init" until we receive the first request */
+	tctx.target = uk_boot_shutdown_barrier();
+#endif /* CONFIG_LIBUKBOOT_MAINTHREAD */
+
+exit:
+	uk_pr_info("Halting system (%d)\n", tctx.target);
+
+	/**
+	 * Call termination functions from init table in reverse order
+	 */
+	/* NOTE: The init loop left `init_entry` at the position that is one
+	 *       step further after the last successfully initialized  entry
+	 */
+	init_entry--;
+	uk_inittab_foreach_reverse2(init_entry, uk_inittab_start) {
+		UK_ASSERT(init_entry);
+
+		if (!init_entry->term)
+			continue;
+
+		uk_pr_debug("Call term function: %p(%p)...\n",
+			    init_entry->term, &tctx);
+		(*init_entry->term)(&tctx);
+	}
+
+	ukplat_terminate(tctx.target); /* does not return */
+}
+
+static inline int do_main(int argc, char *argv[])
+{
+#if CONFIG_LIBPOSIX_ENVIRON
+	char **envp;
```

Comment:
Aw man, it was almost an xmas tree style local variable declaration :cry:, maybe we can make it like that and slowly turn the codebase to follow that aesthetic pattern :pray:.

Reply:
Since when are we following christmas tree order of variable declarations, I usually group them into logical connection and then apply trees per group... I can update this one just for you 😄